provider/terraform/resources/resource_storage_default_object_acl.go

```diff
@@ -2,185 +2,109 @@ package google
-				MinItems: 1,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateRoleEntityPair,
+				},
 			},
 		},
 	}
 }
 
-func resourceStorageDefaultObjectAclCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceStorageDefaultObjectAclCreateUpdate(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
 
 	bucket := d.Get("bucket").(string)
-	roleEntity := d.Get("role_entity").([]interface{})
-
-	for _, v := range roleEntity {
-		pair, err := getRoleEntityPair(v.(string))
-
-		ObjectAccessControl := &storage.ObjectAccessControl{
+	defaultObjectAcl := []*storage.ObjectAccessControl{}
+	for _, v := range d.Get("role_entity").(*schema.Set).List() {
+		pair := getValidatedRoleEntityPair(v.(string))
+		defaultObjectAcl = append(defaultObjectAcl, &storage.ObjectAccessControl{
 			Role:   pair.Role,
 			Entity: pair.Entity,
-		}
-
-		log.Printf("[DEBUG]: setting role = %s, entity = %s on bucket %s", pair.Role, pair.Entity, bucket)
-
-		_, err = config.clientStorage.DefaultObjectAccessControls.Insert(bucket, ObjectAccessControl).Do()
-
-		if err != nil {
-			return fmt.Errorf("Error setting Default Object ACL for %s on bucket %s: %v", pair.Entity, bucket, err)
-		}
+		})
 	}
-	d.SetId(bucket)
-	return resourceStorageDefaultObjectAclRead(d, meta)
-}
-
-func resourceStorageDefaultObjectAclRead(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*Config)
-
-	bucket := d.Get("bucket").(string)
-
-	roleEntities := make([]interface{}, 0)
-	reLocal := d.Get("role_entity").([]interface{})
-	reLocalMap := make(map[string]string)
-	for _, v := range reLocal {
-		res, err := getRoleEntityPair(v.(string))
-
-		if err != nil {
-			return fmt.Errorf(
-				"Old state has malformed Role/Entity pair: %v", err)
-		}
-
-		reLocalMap[res.Entity] = res.Role
-	}
-
-	res, err := config.clientStorage.DefaultObjectAccessControls.List(bucket).Do()
 
+	res, err := config.clientStorage.Buckets.Get(bucket).Do()
 	if err != nil {
-		return handleNotFoundError(err, d, fmt.Sprintf("Storage Default Object ACL for bucket %q", d.Get("bucket").(string)))
+		return fmt.Errorf("Error reading bucket %s: %v", bucket, err)
 	}
 
-	for _, v := range res.Items {
-		role := v.Role
-		entity := v.Entity
-		// We only store updates to the locally defined access controls
-		if _, in := reLocalMap[entity]; in {
-			roleEntities = append(roleEntities, fmt.Sprintf("%s:%s", role, entity))
-			log.Printf("[DEBUG]: saving re %s-%s", v.Role, v.Entity)
+	// Even with ForceSendFields the empty array wasn't working. Luckily, this is the same thing
+	if len(defaultObjectAcl) == 0 {
+		_, err = config.clientStorage.Buckets.Update(bucket, res).PredefinedDefaultObjectAcl("private").Do()
+		if err != nil {
+			return fmt.Errorf("Error updating default object acl to empty for bucket %s: %v", bucket, err)
+		}
+	} else {
+		res.DefaultObjectAcl = defaultObjectAcl
+		_, err = config.clientStorage.Buckets.Update(bucket, res).Do()
```

Comment:
Tiny race condition perhaps? If changes are made the the bucket itself during this update then would this override them?

Reply:
You're right! Looks like we can [use metadatageneration](https://cloud.google.com/storage/docs/generations-preconditions#rmw-scenario) to stop it - fixed.